README.md
# anytv-kue

Kue Helper for setup and cleanup of Kue

# Changes to be made to old codes

1. Replace all `require('kue');` with
1. **NOTE:** Define namespace when calling createQueue to ensure properly scheduling for multiple applications. ex

    ```javascript
        const kue = require('anytv-kue')();
        const queue = kue.createQueue({
            prefix: 'namespace'
        });
    ```

# Added features
- Setup kue

    ```javascript
    // server.js (before)
    const kue = require('anytv-kue')();
    const queue = kue.createQueue();

    function start () {
        /* server code */
    }

    queue.on('error', err => {
        winston.log('error', 'QUEUE ERROR:', err);
    });

    process.once('SIGTERM', sig => {
        winston.log('SIGTERM', sig);
        queue.shutdown(5000, err => {
            winston.log('error', 'Kue shutdown:', err );
            process.exit(0);
        });
    });

    queue.active((err, ids) => {
        ids.forEach(id => {
            kue.Job.get(id, (_err, job) => {
                job.inactive();
            });
        });
    });

    queue.inactive((err, ids) => {
        ids.forEach(id => {
            kue.Job.get(id, (_err, job) => {
                job.inactive();
            });
        });
    });

    start();

    ```
    ```javascript
    // server.js (now)
    const kue = require('antv-kue')();
    const queue = kue.createQueue();

    function start () {
        /* server code */
    }

    kue.setup(queue);

    start();

    ```
- Activate UI
    ```javascript
        const kue = require('anytv-kue')();
        const queue = kue.createQueue({remove_on_complete:false});
        const express = require('express');
        const app = express();

        //activates UI without auth in `/kue`
        kue.activateUI(app)();
        //activates UI in route `/kueapp` without auth
        kue.activateUI(app)('/kueapp');
        //activates UI with basic auth in `/kue`
        kue.activateUI(app, 'username', 'password')();
        //activates UI with basic auth in `/kueapp`
        kue.activateUI(app, 'username', 'password')('/kueapp');
        //activates UI with custom middleware in `/kue`
        kue.activateUI(app, middleWare)();
        //activates UI with custom middleware in `/kueapp`
        kue.activateUI(app, middleWare)('/kueapp');
    ```
- Default Title
    ```javascript
        const kue = require('anytv-kue')();
        const queue = kue.createQueue();

        queue.create('jobtitle', { test: 123 })
            .save(); //default title will be "{ test: 123 }"

        queue.create('jobtitle', { test: 123, title: '123'})
            .save(); //title will be "123"

        queue.create('jobtitle')
            .save(); //title will be "undefined"
    ```
- Cleanup jobs

    ```javascript
        kue.cleanup(job_type, status);
    ```

- Remove jobs on complete
    ```javascript
      // before
      const kue = require('kue');
      const queue = kue.createQueue();

      queue.create('name', {})
        .removeOnComplete(true)
        .save();

      queue.create('name2', {})
        .removeOnComplete(true)
        .save();
    ```
    ```javascript
      //now
      const kue = require('anytv-kue')({shutdownTimer: 10000});
      const queue = kue.createQueue({remove_on_complete: true});

      queue.create('name', {})
        .save();

      queue.create('name2', {})
        .save();
    ```
    
- `fixed_doubling` and `delay_doubling` custom backoffs
    - fixed_doubling - starts at 2 minutes, then 4, then 8, etc
    - delay_doubling - starts at initial_delay * 2 (2 seconds if no initial delay)
    
    ```javascript
        const kue = require('anytv-kue')();
        const queue = kue.createQueue();
  
        queue.createJob('jobtitle', { test: 123 })
            .backoff('fixed_doubling')
            .save(); // next delay will be 2 minutes
          
        queue.createJob('jobtitle', { test: 123 })
            .delay(1000 * 10) // 10 seconds
            .backoff('delay_doubling')
            .save(); // next delay will be 20 seconds
          
        queue.createJob('jobtitle', { test: 123 })
            .backoff('delay_doubling')
            .save(); // next delay will be 2 seconds
    ```

# Available configurations

## constructor
- `shutdownTimer` - time alotted for graceful shutdown

## options
- `remove_on_complete` - will Kue remove data from redis when job is complete
- for other options, check out: https://github.com/lykmapipo/kue-scheduler#options

## setup

```javascript
kue.setup(queue, callbacks);
```

- `queue` - the queue we want to listen to (usually the return of `kue.createQueue`)
- `callbacks` - there are 5 callbacks available
    ```javascript
    {
      //called when there is an error in redis queue
      error: function (err) {

      },

      //called when a job failed altogethere after alotted attempts
      failed: function (jobid) {

      },

      //called when a process SIGTERM occurs
      sigterm: function (sig) {

      },

      //called upon setup, control what you want to do with active jobs
      active: function (err, ids) {

      },

      //called upon setup, control what you want to do with inactive jobs
      inactive: function (err, ids) {

      }
    }
    ```

#How To's

## Update Delayed jobs
```javascript
    'use strict';

    const kue = require('kue');
    const queue = kue.createQueue();

    queue.delayed(function (err, selectedJobs) {
        selectedJobs.forEach(function (id) {
            kue.Job.get(id, function (err, job) {
                if (!job.removeOnComplete()) {
                    job.removeOnComplete(true).update();
                }
            });
        });
    });
```
    
`.update()` on a `Job` object will update it. this is an undocumented of [Kue](http://automattic.github.io/kue/)

# Contributing Guidelines
Please read [CONTRIBUTING.md](CONTRIBUTING.md)


CHANGES.md
Version 1.0.1:
 - HD-112618 Upload.tm > Jobs stuck in Complete and Failed [(#9)](https://github.com/anyTV/anytv-node-kue/pull/9)



test/AnyTVKue.spec.js
'use strict';

const sinon = require('sinon');

const AnyTVKue = require('../lib/AnyTVKue').default;
const sandbox = sinon.createSandbox();

/**
 * @test {AnyTVKue}
 */
describe('#AnyTVKue', function () {

    afterEach(function () {
        sandbox.restore();
    });

    /**
     * @test {AnyTVKue#constructor}
     */
    it('creates an instance of AnyTVKue', function () {

        const inst = new AnyTVKue();

        inst.should.be.an.instanceOf(AnyTVKue);
    });

    /**
     * @test {AnyTVKue#createQueue}
     */
    describe('#createQueue', function () {

        it('creates custom kue.create function', function () {

            const inst = new AnyTVKue();
            const queue = inst.createQueue();

            queue.create.isCustom.should.equal(true);
        });

        it('removes jobs on complete by default', function () {

            const inst = new AnyTVKue();
            const queue = inst.createQueue();
            const job = {
                type: 'test_job',
                data: {
                    a: 1
                }
            };

            const testJob = queue.create(job.type, job.data);
            const stub = sandbox.stub(testJob, 'removeOnComplete');
            testJob.save();

            stub.calledWith(true).should.equal(true);
        });

        // https://github.com/anyTV/anytv-node-kue/issues/2
        // https://freedom.myjetbrains.com/youtrack/issue/Delta-62
        it('does not remove completed jobs when remove_on_complete is set to false', function () {

            const inst = new AnyTVKue({
                remove_on_complete: false
            });
            const queue = inst.createQueue();
            const job = {
                type: 'test_job',
                data: {
                    a: 1
                }
            };

            const testJob = queue.create(job.type, job.data);
            const stub = sandbox.stub(testJob, 'removeOnComplete');
            testJob.save();

            stub.calledWith(false).should.equal(true);

            queue.cleanup('test_job');
        });

        describe('#createJob', function () {

            it('jobs should support a fixed custom doubling backoff that starts at 2 minutes', function () {
                const inst = new AnyTVKue();
                const queue = inst.createQueue();
                const data = {
                    a: 1
                };

                const testJob = queue.createJob('test_job', data)
                    .backoff('fixed_doubling');

                testJob._backoff(1).should.equal(1000 * 60 * 2);
            });

            it('jobs should support a custom doubling backoff based on previous attempt\'s delay', function () {
                const inst = new AnyTVKue();
                const queue = inst.createQueue();
                const data = {
                    a: 1
                };

                const testJob = queue.createJob('test_job', data)
                    .backoff('delay_doubling');

                eval(`testJob._backoff = ${testJob._backoff};`); // convert string to function
                testJob._backoff(1, 2000).should.equal(4000);
            });

            it('custom delay-based doubling backoff should default to 2 seconds when no initial delay', function () {
                const inst = new AnyTVKue();
                const queue = inst.createQueue();
                const data = {
                    a: 1
                };

                const testJob = queue.createJob('test_job', data)
                    .backoff('delay_doubling');

                eval(`testJob._backoff = ${testJob._backoff};`); // convert string to function
                testJob._backoff(1, 0).should.equal(2000);
            });

        });

    });

    /**
     * @test {AnyTVKue#activateUI}
     */
    describe('#activateUI', function () {

        it('should throw error when called without the required arguments', function () {

            const inst = new AnyTVKue();

            inst.activateUI.bind(inst, undefined).should.throw();
            inst.activateUI.bind(inst, null).should.throw();
        });

        it('should throw error when called with invalid argument', function () {

            const inst = new AnyTVKue();

            inst.activateUI.bind(inst, {}).should.throw();
        });
    });

    /**
     * @test {AnyTVKue#setup}
     */
    describe('#setup', function () {

        it('should bind error listeners', function () {

            const inst = new AnyTVKue();
            const queue = inst.createQueue();

            const errorSpy = sandbox.spy();
            const failedSpy = sandbox.spy();

            inst.setup(queue, {
                error: errorSpy,
                failed: failedSpy
            });

            queue.emit('error');
            errorSpy.calledOnce.should.equal(true);

            queue.emit('job failed');
            failedSpy.calledOnce.should.equal(true);
        });
    });
});


src/AnyTVKue.js
'use strict';

import _ from 'lodash';
import winston from 'winston';
import baseConfig from './config';
import kue from 'kue-scheduler';
import basic_auth from 'basic-auth-connect';

/**
 * AnyTVKue class injects additional helper functions on kue-scheduler
 * @example
 * const kue = require('anytv-kue')(config.app.KUE);
 * const Queue = kue.createQueue(config.app.KUE_OPTIONS);
 * kue.setup(Queue); // adds error listeners and requeue previous jobs
 */
class AnyTVKue {

    /**
     * Creates an instance of AnyTVKue class
     * @param {object} [config] Config object that setups the Queue object
     */
    constructor(config) {
        _.merge(baseConfig, config || {});
        _.defaults(this, kue);
    }

    /**
     * Creates a Queue object with specific options
     * @param {object} [options] Options to be passed to Queue object
     * @return {object} Queue instance
     */
    createQueue(options) {
        const self = this;
        options = options || {};

        this.queue = kue.createQueue(options);

        if (this.queue.create.isCustom) {
            return this.queue;
        }

        this.queue._create = this.queue.create;

        this.queue.create = function () {
            const params = arguments[1] || {};

            if (!params.title) {
                params.title = JSON.stringify(params);
                arguments[1] = params;
            }
            const createResult = self.queue._create.apply(this, arguments);

            createResult._save = createResult.save;

            createResult.save = function () {
                const remove_on_complete = 'remove_on_complete' in options
                    ? options.remove_on_complete
                    : baseConfig.remove_on_complete;
                createResult.removeOnComplete(remove_on_complete);
                createResult._save.apply(this, arguments);
            };

            return createResult;
        };

        this.queue.create.isCustom = true;

        /**
         * Patch created job's backoff function to support custom backoff types
         */
        let _createJob = this.queue.createJob;

        this.queue.createJob = function (name, payload) {

            let job = _createJob.call(this, name, payload);

            job.backoff = function (backoff_type) {

                // default implementation is a getter when no arguments passed
                // added for backwards compatibility
                if (arguments.length === 0) {
                    return job._backoff;
                }

                let backoff;
                switch (backoff_type) {

                    /**
                     * Custom callback functions
                     * Executed by eval() so cannot contain variables i.e. can't use config
                     * https://github.com/Automattic/kue#failure-backoff
                     */

                    // 2 mins, 4 mins, 8m mins...
                    case 'fixed_doubling':
                        backoff = attempts => Math.pow(2, attempts) * 1000 * 60;
                        break;

                    // custom exponential based on initial delay
                    // starts at 2 seconds when there's no initial delay
                    case 'delay_doubling':
                        backoff = `(attempts, delay) => delay ? delay * 2 : ${baseConfig.default_delay};`;
                        break;

                    default:
                        backoff = backoff_type;
                        break;
                }

                job._backoff = backoff;
                return job;
            };

            return job;
        };

        return this.queue;
    }

    /**
     * Creates a function that can be used for activating KUE UI app
     * @param {object} app Express server
     * @returns {function} function
     * @example
     * activateUI(app, username, password)(route)
     * @example
     * activateUI(app, authMiddleWare)(route)
     * @example
     * activateUI(app)(route)
     */
    activateUI(app) {

        if (!app || !_.isFunction(app.use)) {
            throw new Error('Invalid Argument');
        }

        switch (arguments.length) {
            case 1:
                return (route) => {
                    app.use(route, kue.app);
                };
            case 2:
                return (route) => {
                    route = route || '/kue';

                    app.use(route, arguments[1]);
                    app.use(route, kue.app);
                };
            case 3:
                return (route) => {
                    route = route || '/kue';

                    app.use(route, basic_auth(arguments[1], arguments[2]));
                    app.use(route, kue.app);
                };
        }
    }

    /**
     * Setups event listeners and re-queues previous jobs
     * @param {object} target Queue instance
     * @param {object} [callbacks] Object containing error, failed, sigterm handlers
     */
    setup(target, callbacks) {
        callbacks = callbacks || {};

        target.on('error', callbacks.error || (err => {
            winston.log('error', 'QUEUE ERROR:', err);
        }));

        target.on('job failed', callbacks.failed || ((err) => {
            winston.log('error', 'FAILED JOB:', err);
        }));

        process.once('SIGTERM', callbacks.sigterm || (sig => {
            winston.log('SIGTERM', sig);
            target.shutdown(baseConfig.shutdownTimer || 5000, err => {
                winston.log('error', 'Kue shutdown:', err);
                process.exit(0);
            });
        }));

        const status = ['active', 'inactive'];

        //requeue all active and inactive jobs
        status.forEach(stat => {
            target[stat](callbacks[stat] || ((err, ids) => {
                ids.forEach(id => {
                    kue.Job.get(id, (_err, job) => {
                        if (job) {
                            job.inactive();
                        }
                    });
                });
            }));
        });
    }

    /**
     * Removes all jobs with specified job_name or status
     * @param {string} job_name Job name
     * @param {string} [status='complete'] Job status
     */
    cleanup(job_name, status) {
        kue.Job.rangeByType(job_name, status || 'complete', 0, -1, 'asc', (err, selectedJobs) => {
            if (selectedJobs && selectedJobs.length) {
                selectedJobs.forEach(job => {
                    job.remove();
                });
            }
        });
    }

    /**
     * Returns config
     * @return {object} Config object
     */
    getConfig() {
        return baseConfig;
    }
}

export default AnyTVKue;
